### ISSUE TITLE
Install failed

### ISSUE THREAD

**Issue by gsmetal** [nashby/cityhash#18]
Complie error:

```
$ gem install cityhash
  Building native extensions.  This could take a while...
ERROR:  Error installing cityhash:
    ERROR: Failed to build gem native extension.

    /home/gsmetal/.rvm/rubies/ruby-2.0.0-p0/bin/ruby extconf.rb
creating Makefile

 make
compiling city.cc
cc1plus: warning: command line option ‘-Wdeclaration-after-statement’ is valid for C/ObjC but not for C++ [enabled by default]
cc1plus: warning: command line option ‘-Wimplicit-function-declaration’ is valid for C/ObjC but not for C++ [enabled by default]
city.cc: In function ‘uint32 Hash32Len0to4(const char*, size_t)’:
city.cc:145:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
city.cc: In function ‘void CityHashCrc256Long(const char*, size_t, uint32, uint64*)’:
city.cc:531:5: error: ‘_mm_crc32_u64’ was not declared in this scope
city.cc:540:5: error: ‘_mm_crc32_u64’ was not declared in this scope
city.cc:550:5: error: ‘_mm_crc32_u64’ was not declared in this scope
make: *** [city.o] Error 1
```

Cityhash from googlecode without -msse4.2 compiles successfully.

PS:

```
$ cat /proc/cpuinfo 
processor   : 0
vendor_id   : GenuineIntel
cpu family  : 6
model       : 23
model name  : Intel(R) Core(TM)2 Duo CPU     P8600  @ 2.40GHz
stepping    : 6
microcode   : 0x60c
cpu MHz     : 800.000
cache size  : 3072 KB
physical id : 0
siblings    : 2
core id     : 0
cpu cores   : 2
apicid      : 0
initial apicid  : 0
fdiv_bug    : no
hlt_bug     : no
f00f_bug    : no
coma_bug    : no
fpu     : yes
fpu_exception   : yes
cpuid level : 10
wp      : yes
flags       : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe nx lm constant_tsc arch_perfmon pebs bts aperfmperf pni dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm sse4_1 lahf_lm dtherm tpr_shadow vnmi flexpriority
bogomips    : 4788.61
clflush size    : 64
cache_alignment : 64
address sizes   : 36 bits physical, 48 bits virtual
power management:
```

**Comment by kapilchoudhary** [comment]
@nashby  I am getting same issue when installing this gem. I am using ubuntu 12.04.
 

Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    /home/rails/.rvm/rubies/ruby-2.2.2/bin/ruby extconf.rb
creating Makefile

make "DESTDIR=" clean

make "DESTDIR="
compiling cityhash.cc
compiling city.cc
city.cc: In function ‘uint32 Hash32Len0to4(const char*, size_t)’:
city.cc:145:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
city.cc: In function ‘void CityHashCrc256Long(const char*, size_t, uint32, uint64*)’:
city.cc:531:5: error: ‘_mm_crc32_u64’ was not declared in this scope
city.cc:540:5: error: ‘_mm_crc32_u64’ was not declared in this scope
city.cc:550:5: error: ‘_mm_crc32_u64’ was not declared in this scope
make: *** [city.o] Error 1

make failed, exit code 2

Gem files will remain installed in /home/rails/.rvm/gems/ruby-2.2.2@derek/gems/cityhash-0.8.1 for inspection.
Results logged to /home/rails/.rvm/gems/ruby-2.2.2@derek/extensions/x86-linux/2.2.0/cityhash-0.8.1/gem_make.out

**Comment by vijaychouhan-rails** [comment]
I am also getting the same error while installing it using bundle install

**Gemfile**
gem "cityhash", "0.8.1"

$ gcc --version
gcc (Ubuntu 4.8.4-2ubuntu1~14.04) 4.8.4
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04 LTS
Release:	14.04
Codename:	trusty


**Error trace**
Using ffi 1.9.10
Using childprocess 0.5.6
Installing cityhash 0.8.1 with native extensions

Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    /home/rails/.rvm/rubies/ruby-2.2.2/bin/ruby -r ./siteconf20150911-25746-12dgsbl.rb extconf.rb
creating Makefile

make "DESTDIR=" clean

make "DESTDIR="
compiling city.cc
city.cc: In function ‘uint32 Hash32Len0to4(const char*, size_t)’:
city.cc:145:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   for (int i = 0; i < len; i++) {
                       ^
city.cc: In function ‘void CityHashCrc256Long(const char*, size_t, uint32, uint64*)’:
city.cc:524:31: error: ‘_mm_crc32_u64’ was not declared in this scope
     z = _mm_crc32_u64(z, b + g);                \
                               ^
city.cc:531:5: note: in expansion of macro ‘CHUNK’
     CHUNK(0); PERMUTE3(a, h, c);
     ^
city.cc:524:31: error: ‘_mm_crc32_u64’ was not declared in this scope
     z = _mm_crc32_u64(z, b + g);                \
                               ^
city.cc:540:5: note: in expansion of macro ‘CHUNK’
     CHUNK(29);
     ^
city.cc:524:31: error: ‘_mm_crc32_u64’ was not declared in this scope
     z = _mm_crc32_u64(z, b + g);                \
                               ^
city.cc:550:5: note: in expansion of macro ‘CHUNK’
     CHUNK(33);
     ^
make: *** [city.o] Error 1

make failed, exit code 2

Gem files will remain installed in /home/rails/.rvm/gems/ruby-2.2.2@lexspot/gems/cityhash-0.8.1 for inspection.
Results logged to /home/rails/.rvm/gems/ruby-2.2.2@lexspot/extensions/x86-linux/2.2.0/cityhash-0.8.1/gem_make.out
An error occurred while installing cityhash (0.8.1), and Bundler cannot continue.
Make sure that `gem install cityhash -v '0.8.1'` succeeds before bundling.

**Comment by ejholmes** [comment]
FWIW, we also ran into this issue on AWS. We were compiling cityhash within a `docker build` on C4 instances. When we went to run the image on a C3 instance, the Ruby VM crashed with:

```
/usr/local/bundle/gems/cityhash-0.8.1/lib/cityhash.rb:11: [BUG] Illegal instruction at 0x007fee4cc25764
```

Removing `-march=native` fixes the issue: https://github.com/remind101/cityhash/commit/915151cdb5c012de3f3b80c07652432160468dc9.

I'm not sure if this gem is still maintained, but I would suggest that the default CFLAGS don't include `-march=native` so that bundled gems are more portable across machines.

**Comment by nashby** [comment]
@ejholmes thanks for following up. I just merged PR that removed default `-march=native`. But obviously CRC methods wont work on instances without sse4. Any suggestion how we should handle it? (As you can see master branch is failing on travis because of it)

**Comment by ejholmes** [comment]
@nashby good question! We don't actually use the CRC methods, so I never noticed https://github.com/nashby/cityhash/blob/master/ext/cityhash/city.cc#L514 before.

Maybe instead of removing `-march=native`, there could be any easier way to remove it via `--with-cppflags` (I couldn't see a way to get around the previous logic to remove it from CPPFLAGS without actually altering extconf.rb).

**Comment by nashby** [comment]
Hey guys! Yeah, it's been 5 years already, sorry. I just released 0.9.0 version that should compile without CRC methods. Please give it a try!
